src/support/allocators/node_allocator.h

```diff
@@ -0,0 +1,399 @@
+            ::operator delete(p);
+        }
+    }
+
+    /**
+     * Actual size in bytes that is used for one chunk (node allocation)
+     */
+    [[nodiscard]] size_t ChunkSize() const noexcept
+    {
+        return m_chunk_size_bytes;
+    }
+
+    /**
+     * Calculates bytes allocated by the memory resource.
+     */
+    [[nodiscard]] size_t DynamicMemoryUsage() const noexcept
+    {
+        if (m_chunk_size_bytes == 0) {
+            return 0;
+        }
+        size_t alloc_size = (m_block_size_bytes / m_chunk_size_bytes) * m_chunk_size_bytes;
+        return memusage::MallocUsage(alloc_size) * m_allocated_blocks.size() + memusage::DynamicUsage(m_allocated_blocks);
+    }
+
+    /**
+     * Counts number of free entries in the free list. This is an O(n) operation. Mostly for debugging / logging / testing.
+     */
+    [[nodiscard]] size_t NumFreeChunks() const noexcept
+    {
+        size_t length = 0;
+        auto node = m_free_chunks;
+        while (node) {
+            node = static_cast<ChunkNode const*>(node)->next;
+            ++length;
+        }
+        return length;
+    }
+
+    /**
+     * Number of memory blocks that have been allocated
+     */
+    [[nodiscard]] size_t NumBlocks() const noexcept
+    {
+        return m_allocated_blocks.size();
+    }
+
+    /**
+     * Calculates the required chunk size for the given type.
+     * The memory block needs to be correctly aligned and large enough to hold both T and ChunkNode.
+     */
+    template <typename T>
+    [[nodiscard]] static constexpr size_t CalcRequiredChunkSizeBytes() noexcept
+    {
+        auto alignment_max = std::max(std::alignment_of_v<T>, std::alignment_of_v<ChunkNode>);
+        auto size_max = std::max(sizeof(T), sizeof(ChunkNode));
+
+        // find closest multiple of alignment_max that holds size_max
+        return ((size_max + alignment_max - 1U) / alignment_max) * alignment_max;
+    }
+
+private:
+    /**
+     * Allocate one full memory block which is used to carve out chunks.
+     * The block size is the multiple of m_chunk_size_bytes that comes closest to m_block_size_bytes.
+     */
+    void AllocateNewBlock()
+    {
+        static_assert(sizeof(char) == 1U);
+
+        auto const num_chunks = m_block_size_bytes / m_chunk_size_bytes;
+        m_untouched_memory_iterator = ::operator new(num_chunks* m_chunk_size_bytes);
+        m_untouched_memory_end = static_cast<char*>(m_untouched_memory_iterator) + num_chunks * m_chunk_size_bytes;
+        m_allocated_blocks.push_back(m_untouched_memory_iterator);
+    }
+
+    //! A single linked list of all data available in the pool. This list is used for allocations of single elements.
+    void* m_free_chunks = nullptr;
+
+    //! A contains all allocated blocks of memory, used to free the data in the destructor.
+    std::vector<void*> m_allocated_blocks{};
+
+    //! The pool's size for the memory blocks. First call to Allocate() determines the used size.
```

Comment:
I think this comment is incorrect (this is the size of one chunk; blocks are a multiple of that).